tests/test_discriminated_union.py

```diff
@@ -1232,3 +1232,72 @@ class TestModel(BaseModel):
         'title': 'TestModel',
         'type': 'object',
     }
+
+
+def test_field_validator_wrapping_union() -> None:
+    """https://github.com/pydantic/pydantic/issues/6525"""
+
+    class CatModel(BaseModel):
+        name: Literal['kitty', 'cat']
+
+        @field_validator('name', mode='after')
+        def replace_name(cls, v: str) -> str:
+            return 'cat'
+
+    class SparrowModel(BaseModel):
+        name: Literal['sparrow', 'birdie']
+
+        @field_validator('name', mode='after')
+        def replace_name(cls, v: str) -> str:
+            return 'sparrow'
+
+    Animal = Annotated[
+        CatModel | SparrowModel,
+        Field(discriminator='name'),
+    ]
+
+    ta = TypeAdapter(Animal)
+
+    assert ta.validate_python({'name': 'kitty'}) == CatModel(name='cat')
```

Comment:
What if both models have a common value – I would expect the firstmost defined object to return. Worth testing?

Reply:
Yes indeed! But that's already tested elsewhere (in this file, I can look for the right line tomorrow; if it's not us should be!)

Reply:
(To be clear, I believe the actual behavior is to error during schema building)